### DISCUSSION TITLE
Using C# 9 outside .NET 5

### DISCUSSION THREAD

**Discussion by johnkors** [dotnet/roslyn#47701] [General] [answered]
Hi,

How is C# 9 tied to .NET 5? Will it be possible to use C# 9 outside .NET 5 if using `Microsoft.Net.Compilers.Toolset` ?

**Comment by PathogenDavid** [comment, marked as answer]
As with C#8 and .NET Core 3.0, [C# 9 is only officially supported in combination with a .NET 5 SDK and a .NET 5 runtime.](https://devblogs.microsoft.com/dotnet/welcome-to-c-9-0/#comment-6107) (Edit: Here's a newer source from documentation: [C# language versioning](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/configure-language-version): "C# 9.0 is supported only on .NET 5 and newer versions.")

If you're going to go the unsupported route, you're best off using the .NET 5 SDK and targeting an older framework (IE: `<TargetFramework>netcoreapp3.1</TargetFramework>` ) and explicitly setting your language version with `<LangVersion>9</LangVersion>`.

[`Microsoft.Net.Compilers.Toolset` is explicitly not intended to support using newer compilers with older toolchains](https://github.com/dotnet/roslyn/blob/c7c2a88a2cb7090fa78203184ffe372a5a098e09/docs/wiki/NuGet-packages.md#microsoftnetcompilerstoolset):

> Using it as a long term solution for providing newer compilers on older MSBuild installations is explicitly not supported. That can and will break on a regular basis.

So I wouldn't recommend using it to accomplish this since you'd now be in a double-unsupported scenario.

**Reply by johnkors** [reply]
Thanks!

**Reply by johnkors** [reply]
Follow up: are there any well known unsupported features when using the .NET 5 SDK, using `netcoreapp3.1` & C# 9 combo?

**Reply by PathogenDavid** [reply]
This is mostly off the top of my head, but here you go:

(Take "Works" with a grain of salt. This is based on my knowledge of the features and in a few cases my limited experience testing this scenario or quick tests. These are not guarantees.)

| Feature | Notes |
|-----|-----|
| [Target-typed new](https://github.com/dotnet/csharplang/issues/100)                                                     | Works
| [Skip locals init](https://github.com/dotnet/csharplang/issues/1738)                                                    | Works (Requires manually defining [`SkipLocalsInitAttribute`](https://github.com/dotnet/runtime/blob/6ed3261974e493a54ec4a97ff281381960cf82fc/src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/SkipLocalsInitAttribute.cs))
| [Lambda discard parameters](https://github.com/dotnet/csharplang/issues/111)                                            | Works
| [Native ints](https://github.com/dotnet/csharplang/issues/435)                                                          | Works (Based on `IntPtr` so [newer stuff there](https://github.com/dotnet/runtime/issues/21943) will be missing.)
| [Attributes on local functions](https://github.com/dotnet/csharplang/issues/1888)                                       | Works
| [Function pointers](https://github.com/dotnet/csharplang/issues/191)                                                    | **Partially works** (Managed function pointers work, unmanaged do not and are gated behind CS8889.)
| [Pattern matching improvements](https://github.com/dotnet/csharplang/issues/2850)                                       | Works
| [Static lambdas](https://github.com/dotnet/csharplang/issues/275)                                                       | Works
| [Records](https://github.com/dotnet/csharplang/issues/39)                                                               | Works (Init-only properties require defining [`IsExternalInit`](https://github.com/dotnet/runtime/blob/6ed3261974e493a54ec4a97ff281381960cf82fc/src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/IsExternalInit.cs))
| [Target-typed conditional](https://github.com/dotnet/csharplang/issues/2460)                                            | Works
| [Covariant](https://github.com/dotnet/csharplang/issues/49) [Returns](https://github.com/dotnet/csharplang/issues/2844) | **NO** Runtime support required, gated behind CS8830
| [Extension GetEnumerator](https://github.com/dotnet/csharplang/issues/3194)                                             | Works
| [Module initializers](https://github.com/dotnet/csharplang/blob/master/proposals/csharp-9.0/module-initializers.md)     | Works (Requires manually defining [`ModuleInitializerAttribute`](https://github.com/dotnet/runtime/blob/6ed3261974e493a54ec4a97ff281381960cf82fc/src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/ModuleInitializerAttribute.cs))
| [Extending Partial](https://github.com/dotnet/csharplang/blob/master/proposals/csharp-9.0/extending-partial-methods.md) | Works
| [Top-level statements](https://github.com/dotnet/csharplang/blob/master/proposals/csharp-9.0/top-level-statements.md)   | Works

**Reply by PathogenDavid** [reply]
Since people found my previous comment helpful and I still find myself linking to it, here's an update for C# 10 features!

(As before, take "Works" with a grain of salt. This is based on my knowledge of the features and brief testing with the .NET 6 RC 1 compiler and .NET Framework 4.8. These are not guarantees, also remember this isn't supported by Microsoft.)

| Feature | Notes |
|-----|-----|
| [Record structs](https://github.com/dotnet/csharplang/issues/4334)                                                 | Works
| [Global Using Directive](https://github.com/dotnet/csharplang/issues/3428)                                         | Works
| [Improved Definite Assignment](https://github.com/dotnet/csharplang/issues/4465)                                   | Works (This is actually tied to the compiler version, not the language version. As such it already works in C# 8 and 9)
| [Constant Interpolated Strings](https://github.com/dotnet/csharplang/issues/2951)                                  | Works
| [Extended property patterns](https://github.com/dotnet/csharplang/issues/4394)                                     | Works
| [Sealed record ToString](https://github.com/dotnet/csharplang/issues/4174)                                         | Works
| [Source Generator V2 APIs](https://github.com/dotnet/roslyn/issues/51257)                                          | Works (This is tied to the compiler version, not the language version)
| [Mix declarations and variables in deconstruction](https://github.com/dotnet/csharplang/issues/125)                | Works
| [Async method builder override](https://github.com/dotnet/csharplang/issues/1407)                                  | Works (Requires manually defining [`AsyncMethodBuilderAttribute`](https://github.com/dotnet/runtime/blob/78593b9e095f974305b2033b465455e458e30267/src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/AsyncMethodBuilderAttribute.cs) and any builders you wish to use)
| [Enhanced `#line` directive](https://github.com/dotnet/csharplang/issues/4747)                                     | Works
| [Lambda improvements](https://github.com/dotnet/csharplang/blob/main/proposals/csharp-10.0/lambda-improvements.md) | Works
| [Interpolated string improvements](https://github.com/dotnet/csharplang/issues/4487)                               | Works (Requires manually defining [`InterpolatedStringHandlerAttribute`](https://github.com/dotnet/runtime/blob/78593b9e095f974305b2033b465455e458e30267/src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/InterpolatedStringHandlerAttribute.cs), [`InterpolatedStringHandlerArgumentAttribute`](https://github.com/dotnet/runtime/blob/78593b9e095f974305b2033b465455e458e30267/src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/InterpolatedStringHandlerArgumentAttribute.cs), and any interpolated string handlers you wish to use*)
| [File-scoped namespace](https://github.com/dotnet/csharplang/issues/137)                                           | Works
| [Parameterless struct constructors](https://github.com/dotnet/csharplang/issues/99)                                | Works (Old runtimes have bug with `Activator.CreateInstance(Type)` and `new()` constraint! See https://github.com/dotnet/runtime/issues/6536) -- `Expression.Default`/`Expression.New` also have problems, see https://github.com/dotnet/runtime/issues/45397
| [Caller expression attribute](https://github.com/dotnet/csharplang/issues/287)                                     | Works (Requires manually defining [`CallerArgumentExpressionAttribute`](https://github.com/dotnet/runtime/blob/78593b9e095f974305b2033b465455e458e30267/src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/CallerArgumentExpressionAttribute.cs))

\* Without the framework providing implementations of `ISpanFormattable` for `System.Int32` and such, it will be very tedious to implement a no/low-allocation handler.

-------------------------

Additionally, here's features which are still in preview for C# 10:

| Feature | Notes |
|-----|-----|
| [Static Abstract Members In Interfaces](https://github.com/dotnet/csharplang/issues/4436)            | **NO** Runtime support required, gated behind CS8919/CS8929
| [Generic attributes](https://github.com/dotnet/csharplang/issues/124)                                | **INADVISABLE** [Older runtime support for generic attributes is VERY spotty](https://github.com/dotnet/csharplang/blob/c1ccfd0d6ad59832a7e7711eeb87461190a53e56/meetings/2017/LDM-2017-02-21.md#generic-attributes), runtime behaviors might change for .NET 7, details below...

Generic attributes can be dangerous to use on older runtimes. To give a few examples:

* Prior to .NET Core 3.0, `GetCustomAttributes` will throw `NotSupportedException` when attempting to reflect generic attributes. (`GetCustomAttributesData` seems to work.)
  * (According to https://github.com/dotnet/roslyn/issues/36285, not even .NET Core 3.0 is safe, sounds like 3.1 is the first "safe" version.)
* https://github.com/dotnet/runtime/issues/56887
  * Additionally, depending on how this is resolved the behavior for .NET might change in .NET 7
* https://github.com/dotnet/runtime/issues/56492 (Fixed in .NET 6)
* [C++/CLI will crash when assemblies which contain generic attributes (even if the type isn't used.)](https://github.com/dotnet/csharplang/blob/c1ccfd0d6ad59832a7e7711eeb87461190a53e56/meetings/2021/LDM-2021-08-30.md#generic-attributes)
* https://github.com/dotnet/runtime/issues/58073 (Scheduled for .NET 7)
* [This feature may still get gated behind a runtime check.](https://github.com/dotnet/runtime/issues/56887#issuecomment-893824472)

**Reply by jrmoreno1** [reply]
You wouldn't happen to have a list of features for C#8 would you?
